aptos-move/block-executor/src/modules_and_scripts.rs

```diff
@@ -0,0 +1,407 @@
+    /// multi-version or non-sync data structures, fetches it from the base view.
+    fn read_module_storage_by_key(
+        &self,
+        key: &T::Key,
+    ) -> VMResult<ModuleStorageRead<ModuleStorageEntry>> {
+        match &self.latest_view {
+            ViewState::Sync(state) => {
+                // If the module read has been previously cached, return early.
+                if let Some(read) = state
+                    .captured_reads
+                    .borrow()
+                    .get_captured_module_storage_read(key)
+                {
+                    return Ok(read.clone());
+                }
+
+                // Otherwise, we need to go to the multi-version data structure to get it, and
+                // record under captured reads.
+                let read = state
+                    .versioned_map
+                    .code_storage()
+                    .module_storage()
+                    .get_or_else(key, self.txn_idx, || {
+                        self.get_base_module_storage_entry(key)
+                    })?;
+                state
+                    .captured_reads
+                    .borrow_mut()
+                    .capture_module_storage_read(key.clone(), read.clone());
+                Ok(read)
+            },
+            ViewState::Unsync(state) => {
+                state.read_set.borrow_mut().module_reads.insert(key.clone());
+                Ok(match state.unsync_map.fetch_module(key) {
+                    // For sequential execution, indices do not matter, but we still return
+                    // them to have uniform interfaces.
+                    Some(entry) => ModuleStorageRead::before_txn_idx(self.txn_idx, entry),
+                    None => match self.get_base_module_storage_entry(key)? {
+                        Some(entry) => ModuleStorageRead::storage_version(entry),
+                        None => ModuleStorageRead::DoesNotExist,
+                    },
+                })
+            },
+        }
+    }
+
+    /// Similar to [LatestView::read_module_storage_by_key], but allows to resolve module
+    /// storage entries based on addresses and names.
+    fn read_module_storage(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<ModuleStorageRead<ModuleStorageEntry>> {
+        let key = T::Key::from_address_and_module_name(address, module_name);
+        self.read_module_storage_by_key(&key)
+    }
+
+    /// Similar to [LatestView::read_module_storage], but in case the module does not exist,
+    /// returns a linker VM error.
+    fn get_existing_module_storage_entry_with_version(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> VMResult<(ModuleVersion, Arc<ModuleStorageEntry>)> {
+        self.read_module_storage(address, module_name)?
+            .into_versioned()
+            .ok_or_else(|| module_linker_error!(address, module_name))
```

Comment:
linker error feels weird, but I guess that's consistent with existing behavior?

Reply:
+1, I think we need Result<Option<V>> interfaces everywhere, and caller decides to map it to linker error or not. This way we also avoid multiple calls to executor.